inference/server/oasst_inference_server/user_chat_repository.py

```diff
@@ -110,6 +111,29 @@ async def initiate_assistant_message(
         self, parent_id: str, work_parameters: inference.WorkParameters
     ) -> models.DbMessage:
         logger.info(f"Adding stub assistant message to {parent_id=}")
+
+        # find and cancel all pending messages by this user
+        pending_msg_query = (
+            sqlmodel.select(models.DbMessage)
+            .where(
+                models.DbMessage.role == "assistant",
+                models.DbMessage.state == inference.MessageState.pending,
+            )
+            .join(models.DbChat)
+            .where(
+                models.DbChat.user_id == self.user_id,
+            )
+        )
+
+        pending_msgs: list[models.DbMessage] = (await self.session.exec(pending_msg_query)).all()
+        for pending_msg in pending_msgs:
+            logger.warning(
+                f"User {self.user_id} has a pending message {pending_msg.id} in chat {pending_msg.chat_id}. Cancelling..."
+            )
+            pending_msg.state = inference.MessageState.cancelled
```

Comment:
does this work without explicit session.add() .. asking because the sqlmodel examples always have this extra `session.add(modified_obj)` call .. I never tested if it is actually necessary.

Reply:
yes, session.add is only necessary if the object is detached from the session previously, but probably doesn't hurt to have it
https://docs.sqlalchemy.org/en/20/orm/session_basics.html#adding-new-or-existing-items